fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java

```diff
@@ -1563,59 +1563,120 @@ public void replayAddPartition(PartitionPersistInfo info) throws DdlException {
+                }
+                PartitionConvertContext context = new PartitionConvertContext();
+                context.setAutoPartitionTable(isAutoPartitionTable);
+                context.setFirstPartitionColumnType(firstPartitionColumn.getType());
+                context.setProperties(properties);
+                context.setTempPartition(clause.isTempPartition());
+                List<SingleRangePartitionDesc> singleRangePartitionDescs = multiRangePartitionDesc.convertToSingle(context);
+                partitionNames.addAll(singleRangePartitionDescs.stream()
+                        .map(SingleRangePartitionDesc::getPartitionName).collect(Collectors.toList()));
             }
-            Range<PartitionKey> partitionRange = null;
-            if (partitionInfo instanceof RangePartitionInfo && partition != null) {
-                partitionRange = ((RangePartitionInfo) partitionInfo).getRange(partition.getId());
+        } else {
+            partitionNames.add(clause.getPartitionName());
+        }
+        List<String> existPartition = Lists.newArrayList();
+        List<String> notExistPartition = Lists.newArrayList();
+
+        for (String partitionName : partitionNames) {
+            if (olapTable.checkPartitionNameExist(partitionName, isTempPartition)) {
+                existPartition.add(partitionName);
+            } else {
+                notExistPartition.add(partitionName);
             }
-            olapTable.dropPartition(db.getId(), partitionName, clause.isForceDrop());
-            if (olapTable instanceof MaterializedView) {
-                MaterializedView mv = (MaterializedView) olapTable;
-                SyncPartitionUtils.dropBaseVersionMeta(mv, partitionName, partitionRange);
+        }
+        if (CollectionUtils.isNotEmpty(notExistPartition)) {
+            if (clause.isSetIfExists()) {
+                LOG.info("drop partition[{}] which does not exist", notExistPartition.toString());
+                if (CollectionUtils.isEmpty(existPartition)) {
+                    return;
+                }
+            } else {
+                ErrorReport.reportDdlException(ErrorCode.ERR_DROP_PARTITION_NON_EXISTENT, notExistPartition.toString());
             }
+        }
+        for (String partitionName : existPartition) {
+            // drop
+            if (isTempPartition) {
+                olapTable.dropTempPartition(partitionName, true);
+            } else {
+                Partition partition = olapTable.getPartition(partitionName);
+                if (!clause.isForceDrop()) {
+                    if (partition != null) {
+                        if (stateMgr.getGlobalTransactionMgr()
+                                .existCommittedTxns(db.getId(), olapTable.getId(), partition.getId())) {
+                            throw new DdlException(
+                                    "There are still some transactions in the COMMITTED state waiting to be completed." +
```

Comment:
It's a good error message.
And, it's recommended to put it into ErrorCode.java if it will be shown to users.
and, you can find proper SQLState in https://www.postgresql.org/docs/15/errcodes-appendix.html.
Maybe, the SQLState for this message could be 01000